File: src/main/org/nlogo/agent/Link3D.java
package org.nlogo.agent;

import org.nlogo.api.AgentException;

public strictfp class Link3D
    extends Link
    implements org.nlogo.api.Link3D {
  public Link3D(World world, Turtle end1, Turtle end2, AgentSet breed) {
    super(world, end1, end2, breed);
  }

  public double z1() {
    return ((Turtle3D) end1).zcor();
  }

  public double z2() {
    return ((Topology3D) world.topology).shortestPathZ(((Turtle3D) end1).zcor(), ((Turtle3D) end2).zcor());
  }

  public double pitch() {
    try {
      return ((Protractor3D) world.protractor()).towardsPitch(end1, end2, true);
    } catch (AgentException e) {
      return 0;
    }
  }
}

File: netlogo-core/src/main/agent/Link3D.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.api.{ AgentException, Link3D => ApiLink3D }

class Link3D(world: World, end1: Turtle, end2: Turtle, breed: AgentSet)
  extends Link(world, end1, end2, breed) with ApiLink3D {

  def z1 = end1.asInstanceOf[Turtle3D].zcor
  def z2 = world.topology.asInstanceOf[Topology3D]
    .shortestPathZ(end1.asInstanceOf[Turtle3D].zcor,
                   end2.asInstanceOf[Turtle3D].zcor)
  def pitch =
    try world.protractor.asInstanceOf[Protractor3D].towardsPitch(end1, end2, true)
    catch { case _: AgentException => 0 }
}
